api/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,309 @@
+	// matches at least one port in the list.
+	// +optional
+	List []AdminNetworkPolicyPort `json:"list,omitempty"`
+}
+
+// AdminNetworkPolicyPort describes a port to select
+type AdminNetworkPolicyPort struct {
+	// The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified, this
+	// field defaults to TCP.
+	// +optional
+	// +kubebuilder:default=TCP
+	Protocol *v1.Protocol `json:"protocol,omitempty"`
+
+	// The port on the given protocol. This can either be a numerical or named
+	// port on a pod. If this field is not provided, this matches no port names and
+	// numbers.
+	// If present, only traffic on the specified protocol AND port will be matched.
+	// +optional
+	Port *intstr.IntOrString `json:"port,omitempty"`
+
+	// If set, indicates that the range of ports from port to endPort, inclusive,
+	// should be allowed by the policy. This field cannot be defined if the port field
+	// is not defined or if the port field is defined as a named (string) port.
+	// The endPort must be equal or greater than port.
+	// +optional
+	EndPort *int32 `json:"endPort,omitempty"`
+}
+
+// AdminNetPolRuleAction string describes the AdminNetworkPolicy action type.
+// +enum
+type AdminNetPolRuleAction string
+
+const (
+	// RuleActionPass enables admins to provide exceptions to ClusterNetworkPolicies and delegate this rule to
+	// K8s NetworkPolicies.
+	AdminNetPolRuleActionPass AdminNetPolRuleAction = "Pass"
+	// RuleActionDeny enables admins to deny specific traffic.
+	AdminNetPolRuleActionDeny AdminNetPolRuleAction = "Deny"
+	// RuleActionAllow enables admins to specifically allow certain traffic.
+	AdminNetPolRuleActionAllow AdminNetPolRuleAction = "Allow"
+)
+
+// AdminNetworkPolicyPeer defines an in-cluster peer to allow traffic to/from.
+// Exactly one of the selector pointers should be set for a given peer. If a
+// consumer observes none of its fields are set, they should assume an unknown
+// option has been specified and fail closed.
+type AdminNetworkPolicyPeer struct {
+	// Namespaces defines a way to select a set of Namespaces
+	// +optional
+	Namespaces *NamespaceSet `json:"namespaces,omitempty"`
+	// NamespacedPods defines a way to select a specific set of pods in
+	// in a set of namespaces
+	// +optional
+	NamespacedPods *NamespaceAndPodSet `json:"namespacedPods,omitempty"`
+}
+
+// NamespaceSet defines a flexible way to select Namespaces in a cluster.
+// Exactly one of the selectors should be set.  If a consumer observes none of
+// its fields are set, they should assume an unknown option has been specified
+// and fail closed.
+type NamespaceSet struct {
+	// Self cannot be "false" when it is set.
+	// If Self is "true" then all pods in the subject's namespace are selected.
+	// +optional
+	Self *bool `json:"self,omitempty"`
+
+	// NotSelf cannot be "false" when it is set.
+	// if NotSelf is "true" then all pods not in the subject's Namespace are selected.
+	// +optional
+	NotSelf *bool `json:"notSelf,omitempty"`
```

Comment:
With some linguistic cleverness you might be able to merge `Self` and `NotSelf` into a single `*bool` value, where one is `&true` and the other is `&false`. Or perhaps merge Self with SameLabels, and NotSelf with NotSameLabels, somehow. (I mean, you could already specify `self: true` as `sameLabels: ["kubernetes.io/metadata.name"]` and `notSelf: true` as `notSameLabels: ["kubernetes.io/metadata.name"]`...)

Reply:
+1 to merging with SameLabels and NotSameLabels

Reply:
actually, it doesn't have to be bool. We can have

```
type NamespaceRelation string

const (
    NamespaceSame    NamespaceRelation = "Same"
    NamespaceNotSame NamespaceRelation = "NotSame"
)

type NamespaceSet struct {
    ...
    Related *NamespaceRelation
    ...
}
```

(This would actually work well with "hierarchical namespaces"... is that still a thing?)

Reply:
Done following Dan's suggestion